src/base/utils/net.cpp

```diff
@@ -38,5 +40,49 @@ namespace Utils
         {
             return !QHostAddress(ip).isNull();
         }
+
+        bool canParseSubnet(const QString &subnetStr)
+        {
+            Subnet subnet;
+            return canParseSubnet(subnetStr, subnet);
+        }
+
+        bool canParseSubnet(const QString &subnetStr, Subnet subnet)
+        {
+            const Subnet invalid = qMakePair(QHostAddress(), -1);
+            subnet = QHostAddress::parseSubnet(subnetStr);
+            return subnet != invalid;
+        }
+
+        bool isLoopbackAddress(const QHostAddress &addr)
+        {
+            return addr == QHostAddress::LocalHost
+                    || addr == QHostAddress::LocalHostIPv6
+                    || addr == QHostAddress(QLatin1String("::ffff:127.0.0.1"));
+        }
+
+        bool isIPInRange(const QHostAddress &addr, const QList<Subnet> &subnets)
+        {
+            QHostAddress protocolEquivalentAddress;
+            bool addrConversionOk = true;
+
+            if (addr.protocol() == QAbstractSocket::IPv4Protocol)
+                // always succeeds
+                protocolEquivalentAddress = QHostAddress(addr.toIPv6Address());
+            else
+                // only succeeds when addr is an ipv4-mapped ipv6 address
+                protocolEquivalentAddress = QHostAddress(addr.toIPv4Address(&addrConversionOk));
+
+            for (const Subnet &subnet : subnets)
+                if (addr.isInSubnet(subnet) || (addrConversionOk && protocolEquivalentAddress.isInSubnet(subnet)))
```

Comment:
can we omit `addrConversionOk`?
Because if `toIPv4Address()` fails, I image it will return an invalid address and there is no harm passing it to `isInSubnet()`.

Reply:
The doc says nothing about returned value in this case, so we can get undefined behavior instead.
>This value is valid if the protocol() is IPv4Protocol, or if the protocol is IPv6Protocol, and the IPv6 address is an IPv4 mapped address. (RFC4291). In those cases, ok will be set to true. Otherwise, it will be set to false.

Reply:
There's no harm but it's a bit inefficient compared to checking `addrConversionOk`. I don't mind the variable but if you feel strongly I'll update it.